sdks/python/apache_beam/io/external/xlang_kafkaio_perf_test.py

```diff
@@ -0,0 +1,133 @@
+from apache_beam.testing.load_tests.load_test_metrics_utils import MeasureTime
+from apache_beam.testing.synthetic_pipeline import SyntheticSource
+from apache_beam.testing.test_pipeline import TestPipeline
+from apache_beam.testing.util import assert_that
+from apache_beam.testing.util import equal_to
+from apache_beam.transforms.util import Reshuffle
+
+WRITE_NAMESPACE = 'write'
+READ_NAMESPACE = 'read'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class KafkaIOTestOptions(LoadTestOptions):
+  @classmethod
+  def _add_argparse_args(cls, parser):
+    parser.add_argument(
+        '--test_class', required=True, help='Test class to run.')
+
+    parser.add_argument('--kafka_topic', required=True, help='Kafka topic.')
+
+    parser.add_argument(
+        '--bootstrap_servers', help='URL TO Kafka Bootstrap service.')
+
+
+class KafkaIOBatchPerfTest:
+  """Performance test for cross-language Kafka IO batch pipeline."""
+  def run(self):
+    write_test = _KafkaIOBatchWritePerfTest()
+    read_test = _KafkaIOBatchReadPerfTest()
+    write_test.run()
+    read_test.run()
+
+
+class _KafkaIOBatchWritePerfTest(LoadTest):
+  def __init__(self):
+    super().__init__(WRITE_NAMESPACE)
+    self.test_options = self.pipeline.get_pipeline_options().view_as(
+        KafkaIOTestOptions)
+    self.kafka_topic = self.test_options.kafka_topic
+    # otherwise see 'ValueError: Unexpected DoFn type: beam:dofn:javasdk:0.1'
+    self.pipeline.not_use_test_runner_api = True
+
+  def test(self):
+    _ = (
+        self.pipeline
+        | 'Generate records' >> iobase.Read(
+            SyntheticSource(self.parse_synthetic_source_options())) \
+            .with_output_types(typing.Tuple[bytes, bytes])
+        | 'Count records' >> beam.ParDo(CountMessages(self.metrics_namespace))
+        | 'Avoid Fusion' >> Reshuffle()
+        | 'Measure time' >> beam.ParDo(MeasureTime(self.metrics_namespace))
+        | 'WriteToKafka' >> kafka.WriteToKafka(
+            producer_config={
+                'bootstrap.servers': self.test_options.bootstrap_servers
+            },
+            topic=self.kafka_topic))
+
+  def cleanup(self):
+    pass
+
+
+class _KafkaIOBatchReadPerfTest(LoadTest):
+  def __init__(self):
+    super().__init__(READ_NAMESPACE)
+    self.test_options = self.pipeline.get_pipeline_options().view_as(
+        KafkaIOTestOptions)
+    self.kafka_topic = self.test_options.kafka_topic
+    # otherwise see 'ValueError: Unexpected DoFn type: beam:dofn:javasdk:0.1'
+    self.pipeline.not_use_test_runner_api = True
+
+  def test(self):
+    output = (
+        self.pipeline
+        | 'ReadFromKafka' >> kafka.ReadFromKafka(
+            consumer_config={
+                'bootstrap.servers': self.test_options.bootstrap_servers,
+                'auto.offset.reset': 'earliest'
+            },
+            topics=[self.kafka_topic],
+            max_num_records=self.input_options['num_records'])
```

Comment:
Do we want this? this will result in using the UnboundedSource, which is the legacy implementation

Reply:
Without setting `max_num_records` the read pipeline will never end until the pipeline timeout if set. The performance test here is actually the python counterpart of [KafkaIOIT.testKafkaIOReadsAndWritesCorrectlyInBatch](https://github.com/apache/beam/blob/fde81f815cc7ddb4943308ffeec4635c7ab360e2/sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java#L231) with the same PTransform parameters.

Do you mean there is a way to read with bounded source now? Then the change could made on both Java and Python performance test as long as the Java and Python tests are still aligned.

Reply:
Sorry, to clarify, this would result in using UnboundedSource as opposed to the SDF implementation, and we don't really care about legacy performance. We need a way to cancel the pipeline while using the SDF implementation

Reply:
Thanks for clarification. I see kafkaio.read.withmaxnumrecords returns a source of `BoundedReadFromUnboundedSource`. Could you please point the place to `SDF implementation` (like which ptransform configuration for kafkaio.read would use this implementation)

Reply:
I see, it is documented here: https://github.com/apache/beam/blob/0917978677f05385eb2ef0d89a016f4a2aa5c6c1/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOReadImplementationCompatibility.java#L54